src/Command_Handler.c

```diff
@@ -0,0 +1,296 @@
+
+#define MAX_PACKET_SIZE (4080)
+#define DEVICE_ID ("Hydrophones v1.0\r\n\0")
+#define MAX_OUTPUT_SIZE (32)
+
+typedef struct Message_Header {
+	uint32_t CRC32;
+	char command[2];
+	uint8_t packet_count;
+	uint8_t packet_idx;
+	uint16_t packet_size;
+	uint16_t total_data_size;
+} Message_Header_t;
+
+static int asciiToInt(char input[], uint8_t length) {
+
+	int output = 0;
+
+	for(int i = 0; i < length; i++) {
+
+		if (input[i] >= '0' && input[i] <= '9') {
+			output = output * 10;
+			output += (int)(input[i] - '0');
+		} else {
+			return -1; //Error: ASCII character is not a number
+		}
+	}
+
+	return output;
+}
+
+static inline uint32_t reverse_bit_order(uint32_t input_word) {
+	uint32_t output_word = input_word;
+	output_word = ((output_word >>  1) & 0x55555555) | ((output_word <<  1) & 0xaaaaaaaa);
+	output_word = ((output_word >>  2) & 0x33333333) | ((output_word <<  2) & 0xcccccccc);
+	output_word = ((output_word >>  4) & 0x0f0f0f0f) | ((output_word <<  4) & 0xf0f0f0f0);
+	output_word = ((output_word >>  8) & 0x00ff00ff) | ((output_word <<  8) & 0xff00ff00);
+	output_word = ((output_word >> 16) & 0x0000ffff) | ((output_word << 16) & 0xffff0000);
+	return output_word;
+}
+
+/* Calculates the CRC32 result of the data with 0xFFFFFFFF starting value.
+ * 	Inputs: - data: pointer to the start of the data
+ * 			- data_size: size of data, in bytes
+ * 	Output: - CRC32 result of data
+ */
+static uint32_t CRC32(uint32_t* data, uint16_t data_size) {
+	uint16_t word_size = data_size / 4;
+	uint8_t extra_bytes = data_size % 4;
+
+	/* Reset the CRC unit */
+	CRC->CR = CRC_CR_RESET;
+
+	/* Compute CRC of data */
+	for(int i = 0; i < word_size; ++i) {
+		CRC->DR = *data++;
+	}
+
+	/* Return the result */
+	return CRC->DR;
+}
+
+void Command_Handler() {
+
+	char commandString[MAX_BUFFER_SIZE];
+	char outputString[MAX_OUTPUT_SIZE];
+	Message_Header_t ADC_message_header;
+	uint16_t packet_size = MAX_PACKET_SIZE;
+	uint8_t* ADC_Buffer_ptr = get_ADC_buffer();
+
+	/* Initialize ADC data read message header */
+	strncpy((char*)ADC_message_header.command, "DR", 2);
+	ADC_message_header.total_data_size = get_ADC_buffer_size();
+	ADC_message_header.packet_size = packet_size - (packet_size % 4);
+	ADC_message_header.packet_count = ADC_message_header.total_data_size / ADC_message_header.packet_size;
+	if(ADC_message_header.total_data_size % ADC_message_header.packet_size) {
+		ADC_message_header.packet_count++;
+	}
+	ADC_message_header.packet_idx = 0;
+
+	while(1) {
+		//it's important that this is while, if the task is accidentally awaken it
+		//can't execute without having at least one item the input buffer
+		while(inputBuffer.size == 0){
+
+			//sleeps the task until it is notified by the UART controller
+			ulTaskNotifyTake( pdTRUE, portMAX_DELAY);
+
+		}
+
+		Buffer_pop(&inputBuffer, commandString);
+
+		/* ---- Check for different commands ---- */
+
+		/* Command to return ID of the device */
+		if(strncmp(commandString, "RID", 3) == 0) {

```

Comment:
Can you use strcmp instead of strncmp for commands that don't have arguments? Right now, this will come true even if some garbage data is followed by "NEXT". I'm trying to do this for the Power Board and Motor Controller as well.

Reply:
What about strncmp with n being the size of the buffer? I think that would be safer than just strcmp in case the data isn't null terminated for some reason. The n is just the max number of bytes it will check, it will still stop at a null terminator if there is one.

Reply:
I'm okay with that idea!